flink-runtime/src/main/java/org/apache/flink/runtime/scheduler/metrics/DeploymentStateTimeMetrics.java

```diff
@@ -0,0 +1,168 @@
+ *
+ * <p>These metrics differentiate between batch & streaming use-cases:
+ *
+ * <p>Batch: Measures from the start of the first deployment until the first task has been deployed.
+ * From that point the job is making progress.
+ *
+ * <p>Streaming: Measures from the start of the first deployment until all tasks have been deployed.
+ * From that point on checkpoints can be triggered, and thus progress be made.
+ */
+public class DeploymentStateTimeMetrics
+        implements ExecutionStateUpdateListener, StateTimeMetric, MetricsRegistrar {
+
+    private static final long NOT_STARTED = -1L;
+
+    private final Predicate<Integer> deploymentStartPredicate;
+    private final Predicate<Integer> deploymentEndPredicate;
+    private final MetricOptions.JobStatusMetricsSettings stateTimeMetricsSettings;
+    private final Clock clock;
+
+    // deployment book-keeping
+    private final Set<ExecutionAttemptID> expectedDeployments = new HashSet<>();
+    private int pendingDeployments = 0;
+    private int completedDeployments = 0;
+
+    // metrics state
+    private long deploymentStart = NOT_STARTED;
+    private long deploymentTimeTotal = 0L;
+
+    public DeploymentStateTimeMetrics(
+            JobType semantic, MetricOptions.JobStatusMetricsSettings stateTimeMetricsSettings) {
+        this(semantic, stateTimeMetricsSettings, SystemClock.getInstance());
+    }
+
+    @VisibleForTesting
+    DeploymentStateTimeMetrics(
+            JobType semantic,
+            MetricOptions.JobStatusMetricsSettings stateTimeMetricsSettings,
+            Clock clock) {
+        this.stateTimeMetricsSettings = stateTimeMetricsSettings;
+        this.clock = clock;
+
+        switch (semantic) {
+            case BATCH:
+                deploymentStartPredicate = completedDeployments -> completedDeployments == 0;
+                deploymentEndPredicate = completedDeployments -> completedDeployments > 0;
+                break;
+            default:
+                deploymentStartPredicate = completedDeployments -> true;
+                deploymentEndPredicate =
+                        completedDeployments -> completedDeployments == expectedDeployments.size();
+                break;
+        }
+    }
+
+    @Override
+    public long getCurrentTime() {
+        return deploymentStart == NOT_STARTED
+                ? 0L
+                : Math.max(0, clock.absoluteTimeMillis() - deploymentStart);
+    }
+
+    @Override
+    public long getTotalTime() {
+        return getCurrentTime() + deploymentTimeTotal;
+    }
+
+    @Override
+    public long getBinary() {
+        return deploymentStart == NOT_STARTED ? 0L : 1L;
+    }
+
+    @Override
+    public void registerMetrics(MetricGroup metricGroup) {
+        StateTimeMetric.register(stateTimeMetricsSettings, metricGroup, this, "deploying");
+    }
+
+    @Override
+    public void onStateUpdate(
+            ExecutionAttemptID execution, ExecutionState previousState, ExecutionState newState) {
```

Comment:
🤔 should we have some guards against invalid transitions?

Reply:
I'd rather formalize the state transitions elsewhere than check for them here.

Reply:
We also don't want to fail the job because of a wrong assumption in the metrics.